src/server/routes/signIn.ts

```diff
@@ -405,6 +406,48 @@ router.post(
   }),
 );
 
+/**
+ * If an Okta session exists, this route will re-authenticate the Okta session
+ * and also refresh the IDAPI session concurrently, synchronising the expiry
+ * times of the Okta and IDAPI sessions, before returning the client to the URL
+ * they came from.
+ */
+router.get(
+  '/signin/refresh',
+  handleAsyncErrors(async (req: Request, res: ResponseWithRequestState) => {
+    const { useOkta, returnUrl } = res.locals.queryParams;
+    const oktaSessionCookieId: string | undefined = req.cookies.sid;
+    const identitySessionCookie = req.cookies.SC_GU_U;
+    const identityLastAccessCookie = req.cookies.SC_GU_LA;
+
+    const { defaultReturnUri } = getConfiguration();
+    const redirectUrl = returnUrl || defaultReturnUri;
+
+    // Check if the user has an existing Okta session.
+    if (okta.enabled && useOkta && oktaSessionCookieId) {
+      try {
+        // If the user session is valid, we re-authenticate them, supplying
+        // the SID cookie value to Okta.
+        await getSession(oktaSessionCookieId);
+        return performAuthorizationCodeFlow(req, res);
```

Comment:
Something I completely forgot about during the planning!

On the OAuth callback endpoint (https://github.com/guardian/gateway/blob/main/src/server/routes/oauth.ts#L174-L185), all 3 identity cookies are actually refreshed/set. However only the `GU_U` and `SC_GU_U` cookies should be set from this call when a refresh happens, not the `SC_GU_LA`.

The best way to do this would be to add an optional flag to the `AuthorizationState` (https://github.com/guardian/gateway/blob/main/src/server/lib/okta/openid-connect.ts#L24), e.g. something like `noRefreshSCGULA?: boolean`.

Set this flag as an option in the `performAuthrozationCodeFlow` method (`performAuthorizationCodeFlow(req, res, { noRefreshSCGULA: true });`).

Then refactor the callback endpoint and the `setIDAPICookies` method to also utilize this flag.

It may sound a bit complicated, so give me a shout if this doesn't make sense!

Reply:
@coldlink The fix is in 41d358884957fc6636041e31fcadc97dc482a064 and the test for it in 20e2ce17131e4f52ee402227f62948e1468e42b2, if you don't mind having another look.

Reply:
Great, looks good to me!